- Encapsulate the image stitching code into a library.

RichardTamarisk opened the pull request:
- Usage:
- cd code/fusion_fuc
- mkdir build
- cd build
- cmake ..
- make
- ./main

RichardTamarisk committed:
- Encapsulate the image stitching code into a library.
- Usage: in fucsion_fuc
- mkdir build
- cd build
- cmake ..
- make

JackLau1222 commented on code/fusion_fuc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+# 查找 OpenCV
+find_package(OpenCV REQUIRED)
+# 查找 SDL2
+find_package(SDL2 REQUIRED)
```
in this example, you didn't use the SDL2 library, maybe remove this be better?

JackLau1222 replied:
remove?

JackLau1222 commented on code/fusion_fuc/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+
+# 目标链接 OpenCV 和 SDL2 库
+target_link_libraries(stitcher ${OpenCV_LIBS} SDL2::SDL2)
+target_link_libraries(main stitcher ${OpenCV_LIBS} SDL2::SDL2)
```
and these code line need to be modified

RichardTamarisk replied:
OK

JackLau1222 commented on code/fusion_fuc/include/stitcher.h:
```diff
@@ -0,0 +1,17 @@
+#endif
+
+extern "C" {
+    EXPORT bool stitchImages(const char* img1_path, const char* img2_path, const char* output_path);
```
this interface is not match our desgin, please refer to 
```cpp
   /**
     * Fuses two AVFrames into a single fused frame.
     *
     * This function takes two input AVFrames, frame1 and frame2, and fuses them into a single output frame, frame_fused.
     * The fusion process is performed by averaging the pixel values of the two input frames.
     *
     * @param frame1 The first input AVFrame.
     * @param frame2 The second input AVFrame.
     * @param frame_fused The output AVFrame that will contain the fused image.
     * @return True if the fusion process is successful, false otherwise.
     */
    bool image_fusion(AVFrame *frame1, AVFrame *frame2, AVFrame *frame_fused)
```

JackLau1222 replied:
and maybe you need to follow this [instruction](https://gist.github.com/foowaa/1d296a9dee81c7a2a52f291c95e55680) to convert AVFrame into cv::Mat

RichardTamarisk replied:
I got it that I would solve it in next two days.

RichardTamarisk committed:
* update image_fusion
* add correct_image
* add class FusionOptions

JackLau1222 commented:
i see you add some code into task.hpp, but it is not my mean.

you need to make your code into one module as library so that we can use it in any software.

the bool correct_image(AVFrame *frame_input, AVFrame *frame_output); interface is what i need to use, you need implement it in your code rather than my task.hpp

JackLau1222 commented:
The interface i declared in my code is just show you the interface design

RichardTamarisk commented:
> The interface i declared in my code is just show you the interface design

ok

RichardTamarisk commented:
> i see you add some code into task.hpp, but it is not my mean.
> 
> you need to make your code into one module as library so that we can use it in any software.
> 
> the bool correct_image(AVFrame *frame_input, AVFrame *frame_output); interface is what i need to use, you need implemente it in your code rather than my task.hpp

Ok, I see. I'll fix it as soon as possible.

RichardTamarisk committed:
* Modified fuction: correct_image and image_fusion.
* Add function avframeToCvmat: AVFrame convert to cv::Mat.
* ADD function cvmatToAvframe: cv::Mat convert to AVFrame.

RichardTamarisk committed:
Update CMakeLists.txt

* remove SDL2

RichardTamarisk committed:
Update CMakeLists.txt

* stitcher.cppis compiled into a dynamic library, which is directly linked to main.cpp

JackLau1222 commented on code/fusion_fuc/include/stitcher.h:
```diff
@@ -0,0 +1,28 @@
+#include <libavformat/avformat.h>
+#include <libavcodec/avcodec.h>
+#include <libavutil/channel_layout.h>
+#include <libavutil/opt.h>
```
Declaration missing extern "C" there.

JackLau1222 commented on code/fusion_fuc/include/stitcher.h:
```diff
@@ -0,0 +1,28 @@
+#include <libavutil/frame.h>
+#include <libavutil/imgutils.h>
+#include <libswscale/swscale.h>
+#include <libavutil/samplefmt.h>
```
missing?

JackLau1222 commented on code/fusion_fuc/main.cpp:
```diff
@@ -0,0 +1,48 @@
+#include <iostream>
+#include <opencv2/opencv.hpp>
+#include <libavformat/avformat.h>
+#include <libavcodec/avcodec.h>
```
there?

JackLau1222 commented on code/fusion_fuc/CMakeLists.txt:
```diff
@@ -0,0 +1,45 @@
+
+# 设置项目名称
+project(ImageFusion)
+
```
Please add one line ``set(CMAKE_CXX_STANDARD 17)``, the std::clamp template has been declared since C++17

RichardTamarisk committed:
* Remove SDL2 and set CMAKE_CXX_STANDARD 17 from file CMakeLists.txt.
* Delete duplicate header files.
* This update will compile normally.
* Although AVFrame can be converted to cv::Mat normally, it is not only
  the image information obtained by using cv::imwread() function, so it
does not meet the requirements in the subsequent feature point detection
and cannot run normally: terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.8.1)
/home/richard/Downloads/opencv/modules/features2d/src/sift.dispatch.cpp:512:
error: (-5:Bad argument) image is empty or has incorrect depth (!=CV_8U)
in function 'detectAndCompute'.

JackLau1222 commented:
i see what code lines you changed, i want to express that you need discuss project problem in this PR comments rather than commit message, and this problem you describe clearly, so you need to find a more matching conversion function that can include all the image details you need.

RichardTamarisk committed:
* Correct misuploaded files.

RichardTamarisk commented:
Thanks for your advice, now I can run the program normally after changing other feature point detection operators, but the image is not correctly stitched as expected, and now it is being perfected.

RichardTamarisk committed:
* Update function: image_fusino.

RichardTamarisk commented:
<img width="985" alt="image" src="https://github.com/user-attachments/assets/e90e460b-4ee3-4517-abf2-a09e73904c26">
At present, the stitching effect is also the current progress effect.I will continue to improve the feature.

JackLau1222 commented:
There is one conflict that task.hpp was modified by myself. So you need to resolve it.
And you should close the conversation you resolved before merging.

RichardTamarisk committed:
Merge branch 'main' into develop_by_richard

RichardTamarisk commented:
> There is one conflict that task.hpp was modified by myself. So you need to resolve it.
> And you should close the conversation you resolved before merging.

RichardTamarisk commented:
Now that the code conflict and issue has been resolved, I think we can merge the pr.

JackLau1222 commented on code/video_fusion_player/src/task.hpp:
```diff
@@ -80,20 +86,89 @@ class Task {
+    bool correct_image(AVFrame *frame_input, AVFrame *frame_output) {
+        if (!frame_input || !frame_output) {
+            return false;
+        }
```
the task.hpp has been modified?

RichardTamarisk committed:
* Modefied task.hpp

RichardTamarisk committed:
* Update task.hpp.

JackLau1222 commented on code/video_fusion_player/src/task.hpp:
```diff
@@ -80,6 +80,20 @@ class Task {
+     * @param frame_output The output AVFrame that will contain the corrected and cropped image.
+     * @return True if the correction and cropping process is successful, false otherwise.
+     */
+    // bool correct_image(AVFrame *frame_input, AVFrame *frame_output) {}
```
There are code changes compared with commit before your PR. 
but in latest commit in main branch, these changes already exist.
To resovle this problem, i recommand two methods:
1. rebase on main branch
2. merge main branch again

The rebase method is recommand in real project work. But the merge method is easier, you can select the one you like to resolve this conflict.
